janitor/functions/conditional_join.py

```diff
@@ -1269,30 +1264,55 @@ def _create_frame(
     if set(df.columns).intersection(right.columns):
         df, right = _create_multiindex_column(df, right)
 
-    if how == "inner":
+    def _inner(
+        df: pd.DataFrame,
+        right: pd.DataFrame,
+        left_index: pd.DataFrame,
+        right_index: pd.DataFrame,
+    ) -> pd.DataFrame:
+        """Create DataFrame for inner join"""
         df = {key: value._values[left_index] for key, value in df.items()}
         right = {
             key: value._values[right_index] for key, value in right.items()
         }
-        return pd.DataFrame({**df, **right}, copy=False)
+        df.update(right)
+        return pd.DataFrame(df, copy=False)
 
-    if how == "left":
-        right = {
-            key: value._values[right_index] for key, value in right.items()
-        }
-        right = pd.DataFrame(right, index=left_index, copy=False)
-    else:
-        df = {key: value._values[left_index] for key, value in df.items()}
-        df = pd.DataFrame(df, index=right_index, copy=False)
+    if how == "inner":
+        return _inner(df, right, left_index, right_index)
 
-    df = pd.merge(
-        df,
-        right,
-        left_index=True,
-        right_index=True,
-        how=how,
-        copy=False,
-        sort=False,
-    )
-    df.index = range(len(df))
-    return df
+    if sort_by_appearance | (left_index.size == 0):
+        if how in {"inner", "left"}:
```

Comment:
`how` can't be inner anymore here (if it was, the function already returned above)

Reply:
If sort by appearance how can be inner, the previous function does not assure that the order is sorted

Reply:
I'm not sure I'm following. if `how == 'inner'`, then it would enter L1282, and it can't ever reach L1285, no?

![image](https://user-images.githubusercontent.com/30731072/190910720-797828fc-1cea-47d6-9059-1d15d9cb6568.png)

Reply:
Ahhh I see what U mean ... Lemme fix that